### DISCUSSION TITLE
Allowing Mission Uploads in flight - WIP

### DISCUSSION THREAD

**Discussion by b14ckyy** [iNavFlight/inav#10269] [General]
With new Serial-Link based Control Systems coming and mLRS natively support MSP-RC Control in the next update, the possibility to use a GCS with a direct serial connection is easer than ever. With this of course also Mission Management on the fly (literally) makes a lot of sense and I am working on that right now (and get a bit coding experience on the way). 

I have disabled the blocker that rejects WP Uploads when the craft is armed and replaced it only with a blocker that rejects if the WP Mode is active. This works very well so far and prevents the accidental interruption and automatic WP-RTH from a running mission. 

To update or replace an ongoing mission, the Pilot has to switch into any other flight mode (like LOITER), upload the new mission, and then enable WP mode again. 

Now a small problem I faced and not sure what the best way of action is: 

Because `nav_wp_mission_restart = RESUME` is the default option, this can lead to a weird behavior. If a mission is flown and the plane is at a certain Waypoint Index, lets say targeting WP 10 but then the Pilot pauses, Uploads a mission with just 5 Waypoints, the old WP10 stays in memory as the last active one. So the plane or copter will continue to fly there and then end the mission with a default infinite loiter. This can be a problem especially if that WP is out of comms range. 

I see 2 options to solve that: 
- During the WP Upload we compare the number of total WP in the currently active mission with the last WP Index stored for the RESUME option. If the total number of WP in the new mission is smaller than that Index number, the Mission index is reset to 0. If its its the same or bigger, we leave the resume. 
- We always reset the WP Index no matter what as soon as a new WP is received from the GCS. This would assume that the pilot only wants to upload a new mission if he wants to fly it all the way from the start. 

Personally I have a tendency towards option 1 as it would allow small adjustments of missions like altitudes or WP locations on the fly without starting all over. 

@breadoven you are the Expert for Mission and navigation stuff, any thoughts? 

Opened a PR: https://github.com/iNavFlight/inav/pull/10273

**Comment by MrD-RC** [comment]
How about option 3. Have a flag to reset the current waypoint. That way if the flag is set. The waypoint reset to 0. If not, it continues at the current waypoint. If the number of waypoints are less that the current waypoint, either reset to 0 or set to the last waypoint.

**Reply by b14ckyy** [reply]
So basically Option 1 but additionally a user set flag if its an entirely new mission? For my 2 Options I would know how to do that and can find how to change the code accordingly but no idea how to do that flag thing :D

**Comment by breadoven** [comment]
I'd need to check how the resume mission works exactly but I thought it gets reset when you load a new mission so the new mission always starts at the beginning. Shouldn't make any difference how the mission is loaded, multimission or MSP, but maybe something got missed and it doesn't work correctly.

If you wanted to prevent the reset to start at the beginning you'd have to tell the FC not to reset the resume function since it just sees a mission upload as a new mission, it can't tell if it's a slightly amended version of the current mission obviously. So a flag or some other more complicated method perhaps.

**Reply by b14ckyy** [reply]
I wanna try to make that myself this time and learn about INAV coding a bit more. My Main concern was if you would agree with one of the 2 behavior options. Right now the RESUME flag does not get reset on mission upload in flight. I guess this was never tested as until 7.1.2 mission upload is blocked/rejected while armed. 

So the question is: Do we want to always reset or just reset if the last active WP number is higher than the number of WP of the new mission?
